### PR TITLE
protected routes


### LINKED ISSUE #9: Protected Routes
Add a wrapper around `Route` from `react-router-dom` to prevent users who aren't logged in from visiting certain pages. Also add functionality for role based access, but don't worry about implementing user roles (user roles will be stored in cookies).

See the [WMK example](https://github.com/ctc-uci/waymakers-frontend/blob/master/src/components/ProtectedRoute/ProtectedRoute.js) from last year. `verifyToken` will be completed in a later sprint, so just add a dummy function there for now.

### PR DESCRIPTION
**Pull request by ctc-devops**
Add a wrapper around `Route` from `react-router-dom` to prevent users who aren't logged in from visiting certain pages. Also add functionality for role based access, but don't worry about implementing user roles (user roles will be stored in cookies).

See the [WMK example](https://github.com/ctc-uci/waymakers-frontend/blob/master/src/components/ProtectedRoute/ProtectedRoute.js) from last year. `verifyToken` will be completed in a later sprint, so just add a dummy function there for now.

closes #9

### CODE REVIEW

**Comment by claudeisakeeb** [comment]
This is a pretty barebones implementation.
For "functionality for role based access" should we write some sample code and comment it out or how should we show it?

File: src/App.jsx
Comment (Vakhshoori101) [inline src/App.jsx]: Please format the `ProtectedRoute` component similar to [here](https://github.com/ctc-uci/waymakers-frontend/blob/master/src/index.js). Looks cleaner :)

File: src/utils/ProtectedRoute.jsx
Comment (Vakhshoori101) [inline src/utils/ProtectedRoute.jsx]: spelled 'the' wrong btw

File: src/utils/ProtectedRoute.jsx
Comment (Vakhshoori101) [inline src/utils/ProtectedRoute.jsx:26]: return a `Route` component like [here](https://github.com/ctc-uci/waymakers-frontend/blob/master/src/components/ProtectedRoute/ProtectedRoute.js)

**Comment by ZubairSidhu** [comment]
For role based access, assume that the users role is stored in a cookie named `role`, then pass a prop to `ProtectedRoute` that allows developers to specify which roles should have access using an array of strings. ex:
```py
<ProtectedRoute path="/admin-only" component={AdminOnly} roles={["admin"]} />
<ProtectedRoute path="/user-only" component={UserOnly} roles={["user"]} />
<ProtectedRoute path="/both" component={Both} roles={["admin", "user"]} />
```
You can set this `role` cookie yourself in a few different ways, but [using inspect element](https://superuser.com/a/1114501) might be the easiest. Feel free to take a look at [#14](https://github.com/ctc-uci/authentication-frontend/pull/14) for more context on the `role` cookie.

**Review by Vakhshoori101 — CHANGES_REQUESTED** [review]
pls fix

File: src/App.jsx
Comment (claudeisakeeb) [reply src/App.jsx]: Ever since react-router-dom upgraded to v6 that syntax doesn't work anymore: https://stackoverflow.com/questions/69864165/error-privateroute-is-not-a-route-component-all-component-children-of-rou

File: src/utils/ProtectedRoute.jsx
Comment (claudeisakeeb) [reply src/utils/ProtectedRoute.jsx:26]: Reference comment above

File: src/utils/ProtectedRoute.jsx
Comment (claudeisakeeb) [reply src/utils/ProtectedRoute.jsx]: Oh you haven't heard? Ahh I guess the dictionary hasn't been updated either so I GUESS I'll change it to the "correct" spelling...

**Comment by claudeisakeeb** [comment]
> For role based access, assume that the users role is stored in a cookie named `role`, then pass a prop to `ProtectedRoute` that allows developers to specify which roles should have access using an array of strings. ex:
> 
> ```python
> <ProtectedRoute path="/admin-only" component={AdminOnly} roles={["admin"]} />
> <ProtectedRoute path="/user-only" component={UserOnly} roles={["user"]} />
> <ProtectedRoute path="/both" component={Both} roles={["admin", "user"]} />
> ```
> 
> You can set this `role` cookie yourself in a few different ways, but [using inspect element](https://superuser.com/a/1114501) might be the easiest. Feel free to take a look at [#14](https://github.com/ctc-uci/authentication-frontend/pull/14) for more context on the `role` cookie.

Ok I think we will wait until #14 makes their final changes and gets merged

File: src/App.jsx
Comment (ZubairSidhu) [reply src/App.jsx]: Here is a method that should work with v6, while keeping the cleaner look of v5: [https://stackoverflow.com/a/70866269](https://stackoverflow.com/a/70866269)

Comment (claudeisakeeb) [reply src/App.jsx]: Implemented

File: src/utils/ProtectedRoute.jsx
Comment (claudeisakeeb) [reply src/utils/ProtectedRoute.jsx:26]: Implemented with Zubair's comment above

**Comment by claudeisakeeb** [comment]
I addressed all changes but this code doesn't work without the new ROLE attribute in cookieKeys introduced in #14 (my local file includes it so I could test but it's not pushed since it's part of 14)
Probably best if 14 is merged first and then I can merge everything into here...or I can just pre-emptively merge right now if you guys want.

**Comment by Vakhshoori101** [comment]
#14 has been merged! feel free to update your branch so you get everything to work :)

**Comment by claudeisakeeb** [comment]
Merged

**Review by ZubairSidhu — APPROVED** [review]
LGTM!

**Pull request merged**
